overcooked_ai_py/visualization/extract_events.py

```diff
@@ -0,0 +1,200 @@
+                        held_object_next_t[id_key] = (find_object_in_game_state(state, held_object_next_t["name"], front_pos) or {}).get("object_id")
+                        if held_object_next_t[id_key] is None:
+                            held_object_next_t[id_key] = next(id_generator)
+        return game_states
+    
+    def get_cumulative_events(events, last_timestep):
+        """
+        Receives events for scatter plot and returns events for cumulative plot
+        """
+        def add_cumulative_event(events, s, timestep, action, player=None, item=None):
+            event = {"action": action,
+                    "sum": s,
+                    "timestep": timestep}
+            if player is not None: event["player"] = player
+            if item is not None: event["item"] = item
+            events.append(event)
+
+        cumulative_events = []
+
+        s = 0
+        player_sums = defaultdict(int)
+        a = "pickups_and_drops"
+        
+        # add zero timestep events
+        first_timestep = 0
+        add_cumulative_event(cumulative_events, s, first_timestep, a)
+        for p in set(map(lambda x: x["player"], events)):
+            add_cumulative_event(cumulative_events, s, first_timestep, a, p)
+
+        for e in filter(lambda x:x["action"] in ["pickup", "drop"], events):
+            p = e["player"]
+            t = e["timestep"]
+            s+=1
+            player_sums[p]+=1
+            add_cumulative_event(cumulative_events, s, t, a)
+            add_cumulative_event(cumulative_events, player_sums[p], t, a, p)
+
+        # add cumulative events at last timestep for nice graph
+        add_cumulative_event(cumulative_events, s, last_timestep, a)
+        for p in set(map(lambda x: x["player"], events)):
+            add_cumulative_event(cumulative_events, player_sums[p], last_timestep, a, p)
+        return cumulative_events
+
+    game_states = [state.to_dict() for state in trajectories["ep_states"][traj_idx]]
+    game_states = add_object_ids_to_game_states(game_states)
+    events = []
+    
+    for t, state in enumerate(game_states):
+        for p, player in enumerate(state["players"]):
+            held_object = player["held_object"] or {}
+            if (t+1 < len(game_states)):
+                held_object_next_t = game_states[t+1]["players"][p]["held_object"] or {}
+                if held_object:
+                    if held_object.get("name") != held_object_next_t.get("name"):
+                        add_drop_item_event(events, player, p, t, held_object, terrain)
+                        add_holding_event(events, p, t, held_object)
+                
+                if held_object_next_t and held_object.get("name") != held_object_next_t.get("name"):
+                    add_pickup_item_event(events, player, p, t, held_object_next_t, terrain)
+                    
+            else: #last timestep
+                if held_object:
+                    add_holding_event(events, p, t, held_object)
```

Comment:
This code seems redundant and abstraction-barrier breaking -- we already have code for detecting events (such as picking up objects, dropping them, various soup creation and potting events) in the `overcooked_mdp` class. In fact, the way I would recommend doing this is by parsing the `trajectory['ep_infos']['agent_infos'][<EVENT_NAME>]` dictionaries. You can find a full list of events currently supported in the `EVENT_TYPES` global in `overcooked_mdp.py`. This way, the visualization implementation doesn't depend on the exact implementation details of OvercookedState

Reply:
This would also allow for potentially richer visualization; we distinguish between "useful" and "non-useful" events (i.e. picking up a tomato is non-useful when all pots are full) which gives an extra dimension to the trajectory data

Reply:
I've done small refactor of event logging to be more flexible:
- added event_list that contains all of the events as a list of dicts
    - it's not ideal that we have both ep_game_stats and event_list with similar info, on the other hand, I needed more info than in ep_game_stats alone and did not want to delete ep_game_stats/modify it too much
- Added adjectives of events ("useful", "viable" etc.) into extracted events so in future this can be incorporated into charts (if  it would be immediately useful I can do it before other tasks)
- I've added JSON serialization of NumPy arrays. Implemented it only for ep_infos as I'm not sure what "metadatas" field in trajectory contains

Reply:
I left a comment asking for clarification on why you needed the separate `event_list` in addition to the `ep_game_stats` but it appears that you did that to keep track of object IDs. I think it would be perfectly fine to add that to the `ep_game_stats` dictionary -- as a general rule of thumb, adding new fields to existing dictionaries (like a nested 'object_id' field) is fine as long as you don't remove existing fields.

Reply:
for the sake of this PR, however, we can keep all the trajectory parsing as it is, as I'm hoping that the refactor issue, which involves writing a Trajectory class wrapper over the increasingly messy dictionaries, will clean up a lot of this logic